File: src/lib/export/html.ts
import { highlightCode } from '@/lib/editor/lowlight'
import { resolveCodeLanguage } from '@/lib/editor/code-languages'
import { evaluateMathExpression } from '@/lib/editor/math-js'
import { renderMermaidSource } from '@/lib/editor/mermaid'
import {
  DEFAULT_PAGE_SETUP,
  normalizePageSetup,
  PAPER_SIZES,
  type PageSetup,
} from '@/lib/editor/page-setup'
import {
  buildHeaderFooterLines,
  formatExportDate,
} from '@/lib/editor/page-header-footer'
import { getPageMargins, shouldShowHeaderFooter } from '@/lib/editor/page-segments'
import {
  DOCUMENT_HIGHLIGHT_CSS,
  DOCUMENT_TIPTAP_CSS,
  buildDocumentContentCss,
  buildWatermarkCss,
} from '@/lib/export/document-styles'
import { colorForExport } from '@/lib/export/export-colors'
import {
  extractFontFamiliesFromContentJson,
  googleFontsLinkTags,
  primaryFontFamilyName,
} from '@/lib/editor/google-fonts'

type TipTapNode = {
  type?: string
  text?: string
  marks?: Array<{ type: string; attrs?: Record<string, unknown> }>
  attrs?: Record<string, unknown>
  content?: TipTapNode[]
}

type RenderContext = {
  mermaidSvgBySource: Map<string, string>
}

function escapeHtml(text: string): string {
  return text
    .replaceAll('&', '&amp;')
    .replaceAll('<', '&lt;')
    .replaceAll('>', '&gt;')
    .replaceAll('"', '&quot;')
}

function renderMarks(text: string, marks?: TipTapNode['marks']): string {
  if (!marks?.length) return escapeHtml(text)

  return marks.reduce((acc, mark) => {
    switch (mark.type) {
      case 'bold':
        return `<strong>${acc}</strong>`
      case 'italic':
        return `<em>${acc}</em>`
      case 'underline':
        return `<u>${acc}</u>`
      case 'strike':
        return `<s>${acc}</s>`
      case 'highlight': {
        const color = String(mark.attrs?.color ?? '#fff3a3')
        return `<mark style="background:${color}">${acc}</mark>`
      }
      case 'code':
        return `<code>${acc}</code>`
      case 'link': {
        const href = String(mark.attrs?.href ?? '#')
        return `<a href="${escapeHtml(href)}">${acc}</a>`
      }
      case 'subscript':
        return `<sub>${acc}</sub>`
      case 'superscript':
        return `<sup>${acc}</sup>`
      case 'textStyle': {
        const styles: string[] = []
        if (mark.attrs?.color) styles.push(`color:${colorForExport(String(mark.attrs.color))}`)
        if (mark.attrs?.fontSize) styles.push(`font-size:${mark.attrs.fontSize}`)
        if (mark.attrs?.fontFamily) styles.push(`font-family:${mark.attrs.fontFamily}`)
        if (!styles.length) return acc
        return `<span style="${styles.join(';')}">${acc}</span>`
      }
      default:
        return acc
    }
  }, escapeHtml(text))
}

function calloutColors(variant: string): { accent: string; bg: string; border: string } {
  switch (variant) {
    case 'tip':
      return { accent: '#22c55e', bg: 'rgba(34,197,94,0.1)', border: 'rgba(34,197,94,0.3)' }
    case 'warning':
      return { accent: '#f59e0b', bg: 'rgba(245,158,11,0.12)', border: 'rgba(245,158,11,0.34)' }
    case 'danger':
      return { accent: '#ef4444', bg: 'rgba(239,68,68,0.11)', border: 'rgba(239,68,68,0.32)' }
    default:
      return { accent: '#3b82f6', bg: 'rgba(59,130,246,0.1)', border: 'rgba(59,130,246,0.3)' }
  }
}

function renderInline(nodes: TipTapNode[] | undefined, ctx: RenderContext): string {
  return (nodes ?? [])
    .map((node) => {
      if (node.type === 'text') {
        return renderMarks(node.text ?? '', node.marks)
      }
      if (node.type === 'hardBreak') return '<br />'
      if (node.type === 'wikiLink') {
        const label = String(node.attrs?.label ?? 'document')
        return `<a class="wiki-link" style="color:#007aff;text-decoration:underline;">[[${escapeHtml(label)}]]</a>`
      }
      if (node.type === 'footnote') {
        const number = Number(node.attrs?.number ?? 1)
        const id = String(node.attrs?.id ?? number)
        return `<sup id="fnref-${escapeHtml(id)}"><a href="#fn-${escapeHtml(id)}" style="text-decoration:none;">[${number}]</a></sup>`
      }
      if (node.type === 'mathInline') {
        const expression = String(node.attrs?.expression ?? '')
        const evaluation = evaluateMathExpression(expression)
        const result = evaluation.ok ? ` = ${evaluation.result}` : ''
        return `<span class="math-inline">${escapeHtml(expression)}${escapeHtml(result)}</span>`
      }
      if (node.type === 'emoji') {
        return escapeHtml(String(node.attrs?.name ?? '🙂'))
      }
      return renderNodes(node.content, ctx)
    })
    .join('')
}

function textAlignStyle(attrs?: Record<string, unknown>): string {
  const align = attrs?.textAlign
  if (align === 'center' || align === 'right' || align === 'justify') {
    return ` style="text-align:${align}"`
  }
  return ''
}

function renderMermaidFigure(source: string, ctx: RenderContext): string {
  const svg = ctx.mermaidSvgBySource.get(source.trim())
  if (svg) {
    return `<figure class="mermaid-diagram" style="margin:16pt 0;text-align:center;overflow:auto;">${svg}</figure>`
  }
  return `<figure class="mermaid-diagram" style="margin:16pt 0;"><pre style="white-space:pre-wrap;font-family:ui-monospace,monospace;font-size:10pt;padding:12pt;background:#f5f5f7;border-radius:8pt;">${escapeHtml(source)}</pre></figure>`
}

function renderNodes(nodes: TipTapNode[] | undefined, ctx: RenderContext): string {
  return (nodes ?? [])
    .map((node) => {
      switch (node.type) {
        case 'paragraph':
          return `<p${textAlignStyle(node.attrs)}>${renderInline(node.content, ctx)}</p>`
        case 'heading': {
          const level = Number(node.attrs?.level ?? 1)
          const tag = `h${Math.min(Math.max(level, 1), 6)}`
          return `<${tag}${textAlignStyle(node.attrs)}>${renderInline(node.content, ctx)}</${tag}>`
        }
        case 'bulletList':
          return `<ul>${renderNodes(node.content, ctx)}</ul>`
        case 'orderedList':
          return `<ol>${renderNodes(node.content, ctx)}</ol>`
        case 'listItem':
          return `<li>${renderNodes(node.content, ctx)}</li>`
        case 'blockquote':
          return `<blockquote>${renderNodes(node.content, ctx)}</blockquote>`
        case 'callout': {
          const variant = String(node.attrs?.variant ?? 'info')
          const colors = calloutColors(variant)
          return `<div data-callout data-variant="${escapeHtml(variant)}" style="margin:12pt 0;padding:10pt 12pt;border:1px solid ${colors.border};border-left:4px solid ${colors.accent};border-radius:8pt;background:${colors.bg};">${renderNodes(node.content, ctx)}</div>`
        }
        case 'codeBlock': {
          const raw = (node.content ?? []).map((n) => n.text ?? '').join('')
          const language = node.attrs?.language as string | undefined
          const resolved = resolveCodeLanguage(language)
          const className = resolved ? `hljs language-${resolved}` : 'hljs'
          return `<pre><code class="${className}">${highlightCode(raw, language)}</code></pre>`
        }
        case 'horizontalRule':
          return '<hr />'
        case 'pageBreak':
          return '<div style="page-break-after:always;border-top:1px dashed #ccc;margin:24pt 0;padding-top:8pt;color:#888;font-size:10pt;">Page break</div>'
        case 'details':
          return `<details open style="margin:12pt 0;border:1px solid #ddd;border-radius:8px;padding:8pt 12pt;">${renderNodes(node.content, ctx)}</details>`
        case 'detailsSummary':
          return `<summary style="font-weight:600;cursor:pointer;margin-bottom:8pt;">${renderInline(node.content, ctx)}</summary>`
        case 'detailsContent':
          return renderNodes(node.content, ctx)
        case 'emoji':
          return escapeHtml(String(node.attrs?.name ?? '🙂'))
        case 'youtube': {
          const src = String(node.attrs?.src ?? '')
          return `<div style="margin:16pt 0;aspect-ratio:16/9;"><iframe src="${escapeHtml(src)}" width="100%" height="360" frameborder="0" allowfullscreen></iframe></div>`
        }
        case 'mathInline': {
          const expression = String(node.attrs?.expression ?? '')
          const evaluation = evaluateMathExpression(expression)
          const result = evaluation.ok ? ` = ${evaluation.result}` : ''
          return `<span class="math-inline">${escapeHtml(expression)}${escapeHtml(result)}</span>`
        }
        case 'mathBlock': {
          const expression = String(node.attrs?.expression ?? '')
          const evaluation = evaluateMathExpression(expression)
          const result = evaluation.ok ? ` = ${evaluation.result}` : ''
          return `<div class="math-block" style="margin:12pt 0;text-align:center;font-family:ui-monospace,monospace;">${escapeHtml(expression)}${escapeHtml(result)}</div>`
        }
        case 'mermaidDiagram': {
          const source = String(node.attrs?.source ?? '')
          return renderMermaidFigure(source, ctx)
        }
        case 'table':
          return `<table style="border-collapse:collapse;width:100%;margin:12pt 0;">${renderNodes(node.content, ctx)}</table>`
        case 'tableRow':
          return `<tr>${renderNodes(node.content, ctx)}</tr>`
        case 'tableHeader': {
          const bg = node.attrs?.backgroundColor ? `background:${node.attrs.backgroundColor};` : ''
          return `<th style="border:1px solid #ccc;padding:8pt;text-align:left;${bg}">${renderNodes(node.content, ctx)}</th>`
        }
        case 'tableCell': {
          const bg = node.attrs?.backgroundColor ? `background:${node.attrs.backgroundColor};` : ''
          return `<td style="border:1px solid #ccc;padding:8pt;${bg}">${renderNodes(node.content, ctx)}</td>`
        }
        case 'image': {
          const src = String(node.attrs?.src ?? '')
          const alt = escapeHtml(String(node.attrs?.alt ?? ''))
          const caption = String(node.attrs?.caption ?? '').trim()
          const width = node.attrs?.width
            ? ` width="${node.attrs.width}" style="width:${node.attrs.width}"`
            : ''
          const align = String(node.attrs?.align ?? 'center')
          const textAlign =
            align === 'full'
              ? 'center'
              : align.includes('float')
                ? 'left'
                : align
          const captionHtml = caption
            ? `<figcaption>${escapeHtml(caption)}</figcaption>`
            : ''
          return `<figure data-align="${align}" style="text-align:${textAlign}"><img src="${escapeHtml(src)}" alt="${alt}"${width} />${captionHtml}</figure>`
        }
        case 'taskList':
          return `<ul data-type="taskList">${renderNodes(node.content, ctx)}</ul>`
        case 'taskItem': {
          const checked = node.attrs?.checked ? ' checked' : ''
          return `<li data-type="taskItem"><input type="checkbox"${checked} disabled />${renderNodes(node.content, ctx)}</li>`
        }
        case 'wikiLink':
        case 'footnote':
          return renderInline([node], ctx)
        default:
          return renderNodes(node.content, ctx)
      }
    })
    .join('')
}

type FootnoteRef = { id: string; number: number; content: string }

function collectFootnotes(nodes?: TipTapNode[]): FootnoteRef[] {
  const found: FootnoteRef[] = []

  const walk = (list?: TipTapNode[]) => {
    for (const node of list ?? []) {
      if (node.type === 'footnote') {
        found.push({
          id: String(node.attrs?.id ?? node.attrs?.number ?? found.length + 1),
          number: Number(node.attrs?.number ?? found.length + 1),
          content: String(node.attrs?.content ?? ''),
        })
      }
      walk(node.content)
    }
  }

  walk(nodes)
  return found.sort((a, b) => a.number - b.number)
}

function renderFootnotesSection(footnotes: FootnoteRef[]): string {
  if (!footnotes.length) return ''
  const items = footnotes
    .map(
      (note) =>
        `<li id="fn-${escapeHtml(note.id)}" style="margin:4pt 0;"><span style="font-weight:600;">[${note.number}]</span> ${escapeHtml(note.content)} <a href="#fnref-${escapeHtml(note.id)}" style="text-decoration:none;">↩</a></li>`,
    )
    .join('')
  return `<section class="footnotes" style="margin-top:28pt;padding-top:12pt;border-top:1px solid #ddd;"><h2 style="font-size:12pt;">Footnotes</h2><ol style="padding-left:18pt;font-size:10pt;color:#444;">${items}</ol></section>`
}

function collectMermaidSources(nodes?: TipTapNode[], into = new Set<string>()): Set<string> {
  for (const node of nodes ?? []) {
    if (node.type === 'mermaidDiagram') {
      const source = String(node.attrs?.source ?? '').trim()
      if (source) into.add(source)
    }
    collectMermaidSources(node.content, into)
  }
  return into
}

export async function buildMermaidSvgMap(
  contentJson: string,
  theme: 'neutral' | 'dark' = 'neutral',
): Promise<Map<string, string>> {
  let doc: TipTapNode = { type: 'doc', content: [] }
  try {
    doc = JSON.parse(contentJson) as TipTapNode
  } catch {
    return new Map()
  }

  const sources = [...collectMermaidSources(doc.content)]
  const map = new Map<string, string>()
  await Promise.all(
    sources.map(async (source) => {
      const result = await renderMermaidSource(source, { theme })
      if (result.ok) map.set(source, result.svg)
    }),
  )
  return map
}

export type HtmlExportOptions = {
  pageSetup?: PageSetup
  includeTitleHeading?: boolean
  forPrint?: boolean
  mermaidSvgBySource?: Map<string, string>
}

function buildFirstPageMarginCss(pageSetup: PageSetup): string {
  if (!pageSetup.firstPage.different) return ''

  const first = getPageMargins(pageSetup, 1)
  const standard = getPageMargins(pageSetup, 2)

  if (
    first.top === standard.top &&
    first.right === standard.right &&
    first.bottom === standard.bottom &&
    first.left === standard.left
  ) {
    return ''
  }

  return `
    @page :first {
      margin: ${first.top}px ${first.right}px ${first.bottom}px ${first.left}px;
    }
  `
}

function buildHtmlDocument(
  contentJson: string,
  title: string,
  options: HtmlExportOptions | undefined,
  ctx: RenderContext,
): string {
  const pageSetup = normalizePageSetup(options?.pageSetup ?? DEFAULT_PAGE_SETUP)
  const includeTitleHeading = options?.includeTitleHeading ?? true
  const forPrint = options?.forPrint ?? false
  const paper = PAPER_SIZES[pageSetup.paperSize]
  let doc: TipTapNode = { type: 'doc', content: [] }
  try {
    doc = JSON.parse(contentJson) as TipTapNode
  } catch {
    // fallback empty doc
  }

  const body = renderNodes(doc.content, ctx)
  const footnotes = renderFootnotesSection(collectFootnotes(doc.content))
  const exportDate = formatExportDate()
  const showHeaderFooter = shouldShowHeaderFooter(pageSetup, 1)
  const headerFooter = showHeaderFooter
    ? buildHeaderFooterLines(pageSetup.headerFooter, {
        title,
        page: 1,
        pages: 1,
        date: exportDate,
      })
    : { header: '', footer: '' }

  const watermarkHtml =
    pageSetup.watermark.enabled && pageSetup.watermark.text.trim()
      ? `<div class="export-watermark print-watermark"><span>${escapeHtml(pageSetup.watermark.text.trim())}</span></div>`
      : ''

  const googleFamilies = [
    ...extractFontFamiliesFromContentJson(contentJson),
    primaryFontFamilyName(pageSetup.typography.fontFamily),
  ].filter(Boolean)
  const googleFontLinks = googleFontsLinkTags(googleFamilies)

  return `<!DOCTYPE html>
<html lang="sk">
<head>
  <meta charset="UTF-8" />
  <title>${escapeHtml(title)}</title>
  ${googleFontLinks}
  <style>
    @page {
      size: ${pageSetup.paperSize === 'letter' ? 'letter' : pageSetup.paperSize};
      margin: ${pageSetup.marginTop}px ${pageSetup.marginRight}px ${pageSetup.marginBottom}px ${pageSetup.marginLeft}px;
    }
    ${buildFirstPageMarginCss(pageSetup)}
    html, body {
      margin: 0;
      padding: 0;
      background: ${forPrint ? '#ffffff' : '#f3f4f6'};
    }
    body {
      ${buildDocumentContentCss(pageSetup)}
      max-width: ${paper.width}px;
      margin: 0 auto;
      padding: ${forPrint ? '0' : `${pageSetup.marginTop}px ${pageSetup.marginRight}px ${pageSetup.marginBottom}px ${pageSetup.marginLeft}px`};
      position: relative;
    }
    .document-content {
      position: relative;
      z-index: 1;
      max-width: ${paper.width - pageSetup.marginLeft - pageSetup.marginRight}px;
      margin: 0 auto;
      padding: ${forPrint ? `${pageSetup.marginTop}px ${pageSetup.marginRight}px ${pageSetup.marginBottom}px ${pageSetup.marginLeft}px` : '0'};
    }
    .export-header, .export-footer {
      font-size: 9pt;
      color: #666;
      text-align: center;
    }
    .export-header { margin-bottom: 18pt; padding-bottom: 6pt; border-bottom: 1px solid #ddd; }
    .export-footer { margin-top: 24pt; padding-top: 6pt; border-top: 1px solid #ddd; }
    .mermaid-diagram svg { max-width: 100%; height: auto; }
    ${DOCUMENT_TIPTAP_CSS}
    ${DOCUMENT_HIGHLIGHT_CSS}
    ${buildWatermarkCss(pageSetup.watermark.opacity, pageSetup.watermark.angle)}
    @media print {
      html, body { background: #ffffff; }
      body { padding: 0; }
      .document-content { padding: 0; max-width: none; }
      .export-header {
        position: fixed;
        top: ${Math.max(12, pageSetup.marginTop * 0.35)}px;
        left: ${pageSetup.marginLeft}px;
        right: ${pageSetup.marginRight}px;
      }
      .export-footer {
        position: fixed;
        bottom: ${Math.max(12, pageSetup.marginBottom * 0.35)}px;
        left: ${pageSetup.marginLeft}px;
        right: ${pageSetup.marginRight}px;
      }
    }
  </style>
</head>
<body>
  ${watermarkHtml}
  ${showHeaderFooter && headerFooter.header ? `<div class="export-header">${escapeHtml(headerFooter.header)}</div>` : ''}
  <div class="document-content">
    ${includeTitleHeading ? `<h1>${escapeHtml(title)}</h1>` : ''}
    ${body}
    ${footnotes}
  </div>
  ${showHeaderFooter && headerFooter.footer ? `<div class="export-footer">${escapeHtml(headerFooter.footer)}</div>` : ''}
</body>
</html>`
}

/** Sync HTML export (Mermaid falls back to source `<pre>` unless `mermaidSvgBySource` is provided). */
export function tiptapJsonToHtml(
  contentJson: string,
  title: string,
  options?: HtmlExportOptions,
): string {
  return buildHtmlDocument(contentJson, title, options, {
    mermaidSvgBySource: options?.mermaidSvgBySource ?? new Map(),
  })
}

/** Async HTML export with Mermaid rendered to SVG (for PDF/DOCX/print). */
export async function tiptapJsonToHtmlAsync(
  contentJson: string,
  title: string,
  options?: HtmlExportOptions,
): Promise<string> {
  const mermaidSvgBySource =
    options?.mermaidSvgBySource ?? (await buildMermaidSvgMap(contentJson, 'neutral'))
  return buildHtmlDocument(contentJson, title, options, { mermaidSvgBySource })
}

File: crates/scribe-core/src/html.rs
//! Structural TipTap JSON to HTML rendering.
//!
//! This is the subset of `src/lib/export/html.ts` that needs no browser: text,
//! headings, lists, quotes, code and rules. Rich embeds (Mermaid, D3, maps,
//! media) stay in TypeScript and are skipped here.

use serde_json::Value;

pub fn tiptap_to_html(content_json: &str, title: &str, include_title_heading: bool) -> String {
    let doc: Value = serde_json::from_str(content_json).unwrap_or(Value::Null);

    let mut out = String::from("<article class=\"scribe-export\">");
    if include_title_heading {
        out.push_str("<h1>");
        push_escaped(&mut out, title);
        out.push_str("</h1>");
    }
    render_children(&doc, &mut out);
    out.push_str("</article>");
    out
}

pub fn escape_html(text: &str) -> String {
    let mut out = String::with_capacity(text.len());
    push_escaped(&mut out, text);
    out
}

fn push_escaped(out: &mut String, text: &str) {
    for ch in text.chars() {
        match ch {
            '&' => out.push_str("&amp;"),
            '<' => out.push_str("&lt;"),
            '>' => out.push_str("&gt;"),
            '"' => out.push_str("&quot;"),
            _ => out.push(ch),
        }
    }
}

fn render_children(node: &Value, out: &mut String) {
    let Some(children) = node.get("content").and_then(Value::as_array) else {
        return;
    };
    for child in children {
        render_node(child, out);
    }
}

fn render_node(node: &Value, out: &mut String) {
    match node.get("type").and_then(Value::as_str) {
        Some("text") => render_text(node, out),
        Some("hardBreak") => out.push_str("<br />"),
        Some("paragraph") => {
            out.push_str("<p");
            push_text_align(node, out);
            out.push('>');
            render_children(node, out);
            out.push_str("</p>");
        }
        Some("heading") => {
            let level = node
                .pointer("/attrs/level")
                .and_then(Value::as_i64)
                .unwrap_or(1)
                .clamp(1, 6);
            out.push_str(&format!("<h{level}"));
            push_text_align(node, out);
            out.push('>');
            render_children(node, out);
            out.push_str(&format!("</h{level}>"));
        }
        Some("bulletList") => wrap(node, "<ul>", "</ul>", out),
        Some("orderedList") => {
            let start = node.pointer("/attrs/start").and_then(Value::as_i64);
            match start {
                Some(start) if start != 1 => {
                    out.push_str(&format!("<ol start=\"{start}\">"));
                }
                _ => out.push_str("<ol>"),
            }
            render_children(node, out);
            out.push_str("</ol>");
        }
        Some("listItem") => wrap(node, "<li>", "</li>", out),
        Some("taskList") => wrap(node, "<ul data-type=\"taskList\">", "</ul>", out),
        Some("taskItem") => {
            let checked = node
                .pointer("/attrs/checked")
                .and_then(Value::as_bool)
                .unwrap_or(false);
            out.push_str("<li data-type=\"taskItem\"><input type=\"checkbox\"");
            if checked {
                out.push_str(" checked");
            }
            out.push_str(" disabled />");
            render_children(node, out);
            out.push_str("</li>");
        }
        Some("blockquote") => wrap(node, "<blockquote>", "</blockquote>", out),
        Some("codeBlock") => {
            let language = node
                .pointer("/attrs/language")
                .and_then(Value::as_str)
                .unwrap_or("")
                .trim();
            if language.is_empty() {
                out.push_str("<pre><code>");
            } else {
                out.push_str("<pre><code class=\"language-");
                push_escaped(out, language);
                out.push_str("\">");
            }
            push_escaped(out, &collect_text(node));
            out.push_str("</code></pre>");
        }
        Some("horizontalRule") => out.push_str("<hr />"),
        Some("paintPad") => {
            let ocr = node
                .pointer("/attrs/ocrText")
                .and_then(Value::as_str)
                .unwrap_or("")
                .trim();
            out.push_str("<figure data-type=\"paint-pad\"><p>Paint pad</p>");
            if !ocr.is_empty() {
                out.push_str("<figcaption>");
                push_escaped(out, ocr);
                out.push_str("</figcaption>");
            }
            out.push_str("</figure>");
        }
        // Unknown nodes keep their children so no content is lost.
        _ => render_children(node, out),
    }
}

fn wrap(node: &Value, open: &str, close: &str, out: &mut String) {
    out.push_str(open);
    render_children(node, out);
    out.push_str(close);
}

fn push_text_align(node: &Value, out: &mut String) {
    let align = node
        .pointer("/attrs/textAlign")
        .and_then(Value::as_str)
        .unwrap_or("");
    if matches!(align, "center" | "right" | "justify") {
        out.push_str(&format!(" style=\"text-align:{align}\""));
    }
}

fn render_text(node: &Value, out: &mut String) {
    let text = node.get("text").and_then(Value::as_str).unwrap_or("");
    let marks = node.get("marks").and_then(Value::as_array);

    let Some(marks) = marks.filter(|marks| !marks.is_empty()) else {
        push_escaped(out, text);
        return;
    };

    // Marks wrap the accumulated markup in order, like the TypeScript reducer.
    let mut rendered = escape_html(text);
    for mark in marks {
        rendered = apply_mark(mark, rendered);
    }
    out.push_str(&rendered);
}

fn apply_mark(mark: &Value, inner: String) -> String {
    match mark.get("type").and_then(Value::as_str) {
        Some("bold") => format!("<strong>{inner}</strong>"),
        Some("italic") => format!("<em>{inner}</em>"),
        Some("underline") => format!("<u>{inner}</u>"),
        Some("strike") => format!("<s>{inner}</s>"),
        Some("code") => format!("<code>{inner}</code>"),
        Some("link") => {
            let href = mark
                .pointer("/attrs/href")
                .and_then(Value::as_str)
                .unwrap_or("#");
            format!("<a href=\"{}\">{inner}</a>", escape_html(href))
        }
        Some("highlight") => {
            let color = mark
                .pointer("/attrs/color")
                .and_then(Value::as_str)
                .unwrap_or("#fff3a3");
            format!(
                "<mark style=\"background:{}\">{inner}</mark>",
                escape_html(color)
            )
        }
        Some("textStyle") => {
            let mut styles: Vec<String> = Vec::new();
            if let Some(color) = mark.pointer("/attrs/color").and_then(Value::as_str) {
                styles.push(format!("color:{}", escape_html(color)));
            }
            if let Some(family) = mark.pointer("/attrs/fontFamily").and_then(Value::as_str) {
                styles.push(format!("font-family:{}", escape_html(family)));
            }
            if styles.is_empty() {
                inner
            } else {
                format!("<span style=\"{}\">{inner}</span>", styles.join(";"))
            }
        }
        _ => inner,
    }
}

fn collect_text(node: &Value) -> String {
    if let Some(text) = node.get("text").and_then(Value::as_str) {
        return text.to_string();
    }
    node.get("content")
        .and_then(Value::as_array)
        .map(|children| children.iter().map(collect_text).collect::<String>())
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn renders_small_document() {
        let json = r#"{
            "type": "doc",
            "content": [
                { "type": "heading", "attrs": { "level": 2 }, "content": [{ "type": "text", "text": "Kapitola" }] },
                { "type": "paragraph", "content": [
                    { "type": "text", "text": "Hello " },
                    { "type": "text", "text": "world", "marks": [{ "type": "bold" }, { "type": "italic" }] },
                    { "type": "hardBreak" },
                    { "type": "text", "text": "5 > 3 & \"safe\"" }
                ] },
                { "type": "bulletList", "content": [
                    { "type": "listItem", "content": [{ "type": "paragraph", "content": [{ "type": "text", "text": "item" }] }] }
                ] },
                { "type": "horizontalRule" }
            ]
        }"#;

        let html = tiptap_to_html(json, "Titul", true);

        assert_eq!(
            html,
            concat!(
                "<article class=\"scribe-export\">",
                "<h1>Titul</h1>",
                "<h2>Kapitola</h2>",
                "<p>Hello <em><strong>world</strong></em><br />5 &gt; 3 &amp; &quot;safe&quot;</p>",
                "<ul><li><p>item</p></li></ul>",
                "<hr />",
                "</article>",
            )
        );
    }

    #[test]
    fn omits_title_heading_when_not_requested() {
        let html = tiptap_to_html(r#"{"type":"doc","content":[]}"#, "Titul", false);
        assert_eq!(html, "<article class=\"scribe-export\"></article>");
    }

    #[test]
    fn renders_links_quotes_code_and_tasks() {
        let json = r#"{
            "type": "doc",
            "content": [
                { "type": "blockquote", "content": [{ "type": "paragraph", "content": [
                    { "type": "text", "text": "docs", "marks": [{ "type": "link", "attrs": { "href": "https://a.b?x=1&y=2" } }] }
                ] }] },
                { "type": "codeBlock", "attrs": { "language": "rust" }, "content": [{ "type": "text", "text": "let x = 1 < 2;" }] },
                { "type": "taskList", "content": [
                    { "type": "taskItem", "attrs": { "checked": true }, "content": [{ "type": "paragraph", "content": [{ "type": "text", "text": "done" }] }] },
                    { "type": "taskItem", "attrs": { "checked": false }, "content": [{ "type": "paragraph", "content": [{ "type": "text", "text": "todo" }] }] }
                ] }
            ]
        }"#;

        let html = tiptap_to_html(json, "", false);

        assert!(html.contains(
            "<blockquote><p><a href=\"https://a.b?x=1&amp;y=2\">docs</a></p></blockquote>"
        ));
        assert!(html.contains("<pre><code class=\"language-rust\">let x = 1 &lt; 2;</code></pre>"));
        assert!(html.contains(
            "<li data-type=\"taskItem\"><input type=\"checkbox\" checked disabled /><p>done</p></li>"
        ));
        assert!(html.contains(
            "<li data-type=\"taskItem\"><input type=\"checkbox\" disabled /><p>todo</p></li>"
        ));
    }

    #[test]
    fn renders_text_styles_and_alignment() {
        let json = r##"{
            "type": "doc",
            "content": [
                { "type": "paragraph", "attrs": { "textAlign": "center" }, "content": [
                    { "type": "text", "text": "tinted", "marks": [
                        { "type": "textStyle", "attrs": { "color": "#ff0055", "fontFamily": "Fraunces" } },
                        { "type": "highlight", "attrs": { "color": "#fff3a3" } }
                    ] }
                ] }
            ]
        }"##;

        let html = tiptap_to_html(json, "", false);

        assert!(html.contains("<p style=\"text-align:center\">"));
        assert!(html.contains(
            "<mark style=\"background:#fff3a3\"><span style=\"color:#ff0055;font-family:Fraunces\">tinted</span></mark>"
        ));
    }

    #[test]
    fn unknown_nodes_keep_their_children() {
        let json = r#"{"type":"doc","content":[
            {"type":"callout","content":[{"type":"paragraph","content":[{"type":"text","text":"note"}]}]},
            {"type":"mermaidDiagram","attrs":{"source":"graph TD;"}}
        ]}"#;

        let html = tiptap_to_html(json, "", false);
        assert_eq!(
            html,
            "<article class=\"scribe-export\"><p>note</p></article>"
        );
    }

    #[test]
    fn invalid_json_yields_empty_article() {
        assert_eq!(
            tiptap_to_html("not json", "T", true),
            "<article class=\"scribe-export\"><h1>T</h1></article>"
        );
    }
}
